### PR TITLE
added e2e test case for breadcrumb

### PR DESCRIPTION
**Pull request by NileshPatel17**
issue #796

### CODE REVIEW

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    cy.login()
+    cy.visit(pathWithQueryParams({ path: '/', newParams: { repoUrl: Cypress.env('repoUrl') } }))
+  })
+  it('breadcrumb after navigating to IT Workspace, breadcrumb should be displayed as it should.', () => {
+    const expectedBreadcrumbItems = ['Content', '/', 'IT Workspace']
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="IT Workspace"]').click()
+    cy.get('nav[aria-label="breadcrumb"] li').each(($el) => {
+      expect(expectedBreadcrumbItems).to.include($el.text())
+    })
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:14]: Missing steps:

- Open IT workspace with clicking on it in the tree
- Expected result: children of IT Workspace are listed (Calendar is the first child item in the grid).
- Choose Document library from IT Workspace's children and open it
- Expected result: the breadcrumb contains the following items: Content, IT Worksapce, Document library

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  })
+  it('clicking on parent item in the breadcrumb should open the chosen container', () => {
+    const expectedBreadcrumbItems = ['Content', '/', 'IT Workspace', '/', 'Document Library']
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="IT Workspace"]').click()
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:19]: Missing step:
Expected result: children of IT Workspace are listed. (Calendar is the first child item in the grid).

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    cy.get('[data-test="Document Library"]').click()
+    cy.get('nav[aria-label="breadcrumb"] li').each(($el) => {
+      expect(expectedBreadcrumbItems).to.include($el.text())
+    })
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:23]: This check is not needed here. But you missed some steps:

- Click on IT Workspace in the breadcrumb
- Expected result: children of IT Workspace are listed (Calendar is the first child item in the grid).

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  })
+  it('right click on a breadcrumb item should open its actionmenu.', () => {
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="IT Workspace"]').click()
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:27]: Missing step:
Expected result: children of IT Workspace are listed. (Calendar is the first child item in the grid).

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    cy.get('[data-test="Document Library"]').click()
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:28]:
```suggestion
    cy.get('[data-test="Document Library"]').click({ force: true })
```

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    cy.get('[data-test="IT Workspace"]').click()
+    cy.get('nav[aria-label="breadcrumb"] li').each(($el) => {
+      expect(expectedBreadcrumbItems).to.include($el.text())
+    })
+  })
+  it('clicking on parent item in the breadcrumb should open the chosen container', () => {
+    const expectedBreadcrumbItems = ['Content', '/', 'IT Workspace', '/', 'Document Library']
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="IT Workspace"]').click()
+    cy.get('[data-test="Document Library"]').click()
```
Comment (pusztaienike) [inline apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:20]:
```suggestion
    cy.get('[data-test="Document Library"]').click({ force: true })
```

**Review by pusztaienike — COMMENTED** [review]
The test does not follow the test case. There is no check for the grid, and there are a lot of missing steps. I feel some misunderstanding here.  Let me please illustrate the page with a picture:
image
You missed to check the grid in some cases. I hope that this pic can help a little.

Thanks :)

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts (hunk 1 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:14]: should not "Expected result: children of IT Workspace are listed (Calendar is the first child item in the **grid**)." be "Expected result: children of IT Workspace are listed (Calendar is the first child item in the **list**)?? I can see Calendar in the list, not in the grid. please correct me if  i am wrong.

Comment (pusztaienike) [reply apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:14]: @NileshPatel17,
You have to see them in both places (in the grid and in the list as well). But it is enough to check them in the grid regarding to the test case. See the attached gif.
nileshp

File: apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts (hunk 3 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/breadcrumb/breadcrumb.spec.ts:23]: this is step 4:
**Expected result**: the breadcrumb contains the following items: Content, IT Worksapce, Document library

**Comment by NileshPatel17** [comment]
@pusztaienike changes done.

**Comment by pusztaienike** [comment]
> @pusztaienike changes done.

I have still found some difference. Here is my final version which is working good:

```
import { pathWithQueryParams } from '../../../src/services/query-string-builder'

const expectedBreadcrumbItems = ['Content', '/', 'IT Workspace', '/', 'Document library']

describe('Breadcrumb', () => {
  before(() => {
    cy.login()
    cy.visit(pathWithQueryParams({ path: '/', newParams: { repoUrl: Cypress.env('repoUrl') } }))
  })
  it('breadcrumb after navigating to IT Workspace, breadcrumb should be displayed as it should.', () => {
    cy.get('[data-test="drawer-menu-item-Content"]').click()
    cy.get('[data-test="IT Workspace"]')
      .click()
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('be.visible')
      })

    cy.get('[data-test="Document library"]').click({ force: true })
    cy.get('nav[aria-label="breadcrumb"] li')
      .should('have.length', expectedBreadcrumbItems.length)
      .each(($el) => {
        expect(expectedBreadcrumbItems).to.include($el.text())
      })
  })
  it('clicking on parent item in the breadcrumb should open the chosen container', () => {
    cy.get('[data-test="drawer-menu-item-Content"]').click()
    cy.get('[data-test="IT Workspace"]')
      .click()
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('be.visible')
      })
    cy.get('[data-test="Document library"]')
      .click({ force: true })
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('not.be.visible')
      })
    cy.get('nav[aria-label="breadcrumb"] li')
      .find('button[aria-label="IT Workspace"]')
      .click()
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('be.visible')
      })
  })
  it('right click on a breadcrumb item should open its actionmenu.', () => {
    cy.get('[data-test="drawer-menu-item-Content"]').click()
    cy.get('[data-test="IT Workspace"]')
      .click()
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('be.visible')
      })
    cy.get('[data-test="Document library"]').click({ force: true })
    cy.get('nav[aria-label="breadcrumb"] li')
      .should('have.length', expectedBreadcrumbItems.length)
      .find('button[aria-label="IT Workspace"]')
      .rightclick()
      .then(() => {
        cy.get('ul[role="menu"]').should('be.visible')
      })
  })
})
```
Global:
- I have moved the expectedBreadcrumbItems to upper scope cause I used it from more tests
- *Document Library* has been changed to *Document library*
Test1:
- `.should('have.length', expectedBreadcrumbItems.length)` --> After clicking on Document library the breadcrumb renders more time: the first time it only has ''Content', '/', 'IT Workspace'' but the test won't fail cause the expectedBreadcrumbItems includes these values. You need to check the length as well. In cypress if you have an assertion the test will wait till your condition will be true, so it is a valid solution to wait for rerenders as well. :)
Test2:
- here we don't need to check the breadcrumb, just the grid itself. 
- I also added a plus step here: 
```
    cy.get('[data-test="Document library"]')
      .click({ force: true })
      .then(() => {
        cy.get('[data-test="table-cell-Calendar"]').should('not.be.visible')
      })
```
--> Here I would like to check the clicking on document library causes a rerender of the grid.
Test3:
- assertion added to wait for breadcrumb
- check of the grid has been added

Could you please update your test with my suggestions? 
Thank you for your work! 🥇

**Pull request merged**